nixos/modules/services/x11/xserver.nix

```diff
@@ -155,6 +155,9 @@ in
         "Use services.xserver.fontPath instead of useXFS")
       (mkRemovedOptionModule [ "services" "xserver" "useGlamor" ]
         "Option services.xserver.useGlamor was removed because it is unnecessary. Drivers that uses Glamor will use it automatically.")
+      (mkRemovedOptionModule ["services" "xserver" "videoDriver" ]
+        "Use hardware.gpu.drivers instead")
+      (mkRenamedOptionModule [ "services" "xserver" "videoDrivers" ] [ "hardware" "gpu" "drivers" ])
```

Comment:
I think this option should either stay under xorg or be removed entirely. What's in this list are the xorg DDX drivers, not OGL/VK/whatever drivers. Deriving the latter from the former has always been a bad pattern IMHO.

Instead, we should have an option where you can enable specific drivers directly and ideally also pass driver packages as overrides.

Something like

```
hardware.drivers.<driver> = {
  enable = true;
  package = ...;
  driverSpecificOption = ...;
};
```

for `mesa`, `amdvlk`, `nvidia`, `amdgpu-pro` and whatever other significant driver packages exist.

I also think we might want to have a per-API interface. APIs would include obvious ones like OGL and VK but also OpenCL, VAAPI, oneAPI, CUDA, rocm and various other acceleration APIs for TPUs and the like; anything that facilitates talking to specialised hardware and requires a central userspace component basically.

Each API would have the option to enable/disable (a headless media server might not need OGL/VK but might want VAAPI for instance), declaring available drivers (many APIs allow multiple drivers to co-exist) and perhaps the option to set a preferred/primary driver where applicable.

Something like:

```Nix
hardware.acceleration.<api> = {
  enable = ...;
  packages = { ... };
  preferredPackage = config.hardware.acceleration.packages.<driver>;
  apiSpecificOption = ...;
};
```

The `<driver>.enable` option would place its package into every API the driver implements, allowing the user to override a certain driver's package "globally".

When the desktop is enabled (that setting should also be moved out of xorg), we would enable the `mesa` driver and the OGL and VAAPI APIs.  
If a user requires the nvidia driver, they would set `hardware.drivers.nvidia.enable = true` which would in-turn also enable the kernel module. Should they needed an older driver, they'd set `hardware.drivers.nvidia.package = nvidia_xyz;`.  
Enabling steam (or some other gaming related thing) would enable the VK driver.

What do you think of this design?

Reply:
Your design looks like it resolves my concern over hardware.drivers.packages making mesa hard to override.

What's the benefit of splitting up enabling opengl and vulkan? Bundling everything in mesa together is somewhat expected so the default shouldn't be to split that up or it will add a new confusing way to have games break.

Similarly, mesa should probably default to on if any graphics drivers are in use to match current behavior and behavior of other distros.  
Enabling only nvidia libs should be possible but it's unusual and wouldn't be a good thing to have happen by default or accident.

Reply:
> What's the benefit of splitting up enabling opengl and vulkan?

The reasoning is that these can be en-/disabled separately is that these are entirely separate acceleration APIs. A regular desktop user does not require Vulkan drivers, oneAPI or CUDA.

It also simplifies enabling less usual APIs as all you'd need to do is set `hardware.acceleration.obscureAPI.enable = true` and if there is a "mesa situation" where its the most sensible driver to use, we might opt to enable that driver when the API gets enabled.

Reply:
> A regular desktop user does not require Vulkan drivers

I think this will change soon - wayland compositors are in various stages of investigating or supporting vulkan and I would be surprised if this doesn't eventually become the default.

As long as the defaults are good / the behavior for someone who installs NixOS and wants to have their graphics card work as expected is good then I support your proposal for making very granular options available.